src/terrain/terrain_base.hpp

```diff
@@ -359,18 +482,138 @@ class TerrainBase : public voxel_utility::VoxelBase {
+     * @param y y position
+     * @param guess expected height (for speed)
+     * @return int height
+     */
+    int get_first_not(
+        const std::set<std::pair<int, int>> materials, int x, int y, int guess
+    ) const;
 
-        return ((uint16_t) mat_id << 8) | (uint16_t)color_id;
+    inline uint8_t
+    get_grass_grad_length() const {
+        return grass_grad_length_;
+    }
+
+    inline uint8_t
+    get_grass_mid() const {
+        return grass_mid_;
+    }
+
+    inline std::vector<uint8_t>
+    get_grass_colors() const {
+        return grass_colors_;
+    }
 
-        return TerrainColorMapping::get_colors_inverse_map().at(get_voxel(x, y, z));
+    inline const std::map<uint8_t, const terrain::Material>*
+    get_materials() const {
+        return materials_;
+    }
+
+    /**
+     * @brief initialize area of terrain
+     *
+     * @param area_x area x coordinate
+     * @param area_y area y coordinate
+     * @param gen Generator object that generates tile types
+     */
+    void init_area(int area_x, int area_y, terrain_generation::LandGenerator gen);
+
+    /**
+     * @brief Set a group of tiles
+     *
+     * @param tStamp where the stamp is, and material and color
+     * @param x macro map x position
+     * @param y macro map y position
+     */
+    inline void
+    stamp_tile_region(terrain_generation::TileStamp tStamp, int x, int y) {
+        stamp_tile_region(
+            tStamp.x_start + x * area_size + area_size / 2,
+            tStamp.y_start + y * area_size + area_size / 2, tStamp.z_start,
+            tStamp.x_end + x * area_size + area_size / 2,
+            tStamp.y_end + y * area_size + area_size / 2, tStamp.z_end, tStamp.mat,
+            tStamp.elements_can_stamp, tStamp.color_id
+        );
     }
 
+    /**
+     * @brief Set a group of tiles
+     *
+     * @param x_start lower x position
+     * @param y_start lower y position
+     * @param z_start lower z position
+     * @param x_end greater x position
+     * @param y_end greater y position
+     * @param z_end greater z position
+     * @param mat materials to set
+     * @param color_id color id to set
+     */
+    void stamp_tile_region(
+        int x_start, int y_start, int z_start, int x_end, int y_end, int z_end,
+        const Material* mat, uint8_t color_id
+    );
+    /**
+     * @brief Set a group of tiles
+     *
+     * @param x_start lower x position
+     * @param y_start lower y position
+     * @param z_start lower z position
+     * @param x_end greater x position
+     * @param y_end greater y position
+     * @param z_end greater z position
+     * @param mat materials to set
+     * @param elements_can_stamp type of material that can be changed
+     * @param color_id color id to set
+     */
+    void stamp_tile_region(
+        int x_start, int y_start, int z_start, int x_end, int y_end, int z_end,
+        const Material* mat, std::set<std::pair<int, int>> elements_can_stamp,
+        uint8_t color_id
+    );
```

Comment:
This function is too long

Also make it take a Material& if possible

Also typedefs

Also pass-by-ref